src/drivers/intel/cavs/ssp.c

```diff
@@ -365,10 +366,38 @@ static int ssp_set_config(struct dai *dai,
 		if (ssp_freq[clk_index].enc != CLOCK_SSP_XTAL_OSCILLATOR)
 			sscr0 |= SSCR0_ECS;
 	} else {
-		trace_ssp_error("ssp_set_config() error: BCLK %d",
-				config->ssp.bclk_rate);
-		ret = -EINVAL;
-		goto out;
+		/* try with M/N */
+		for (i = MAX_SSP_FREQ_INDEX; i >= 0; i--) {
+			/* M/N cannot be greater than 1/2 */
+			if ((config->ssp.bclk_rate * 2) > ssp_freq[i].freq)
+				break;
+
+			mn_div = gcd(config->ssp.bclk_rate, ssp_freq[i].freq);
+			i2s_m = config->ssp.bclk_rate / mn_div;
+			i2s_n = ssp_freq[i].freq / mn_div;
+
+			if (i2s_m & (~0xFFFFFF) || i2s_n & (~0xFFFFFF)) {
+				/* M/N values can be up to 24 bits */
+				continue;
+			}
+
+			clk_index = i;
+		}
+
+		if (clk_index < 0) {
+			trace_ssp_error("ssp_set_config() error: BCLK %d",
+					config->ssp.bclk_rate);
+			ret = -EINVAL;
+			goto out;
+		}
+
+		mdivc |= MNDSS(ssp_freq[clk_index].enc);
+		/* division handled by M/N, so mdiv can be 1 */
+		mdiv = 1;
```

Comment:
@lbetlej Can this one affect something in HW? Is it ok to keep it always at 1 or better use it in a way that makes M/N closer to 1/2.? F.e. we can get 6.144MHz BCLK for 24MHz MCLK, with M/N 32/125  & mdiv 1 or with M/N 64/125 & mdiv 2.

Reply:
@lbetlej Well for sure it affects duty cycle,  you have it closer to 50% with higher mdiv.

Reply:
@jajanusz yes, DC is better if closer to 50%, it depends on HW how low DC can be handled
You should find highest divisor and then apply M/N

SCR >= 1 (base div >=1) and SCR + 1 must be even (even divider)

Reply:
@abonislawski Ok, thanks for clarification

Reply:
don't you hav a restriction that M < N/2 or something? IIRC you could not always take mdiv=1

Reply:
@plbossart In spec there is info that M/N has to be <1/2, but it's not true at all. Truth is what we wrote above, that M/(N*mdiv) has to be <1/2 if M/N is enabled. SCR itself (mdiv-1) has to be in range 0 to 4095, but only even dividers give 50% duty cycle, so we should aim for that (no odd mdivs).

Reply:
ok, as long as KarL signs-off on this I am good.